File: sdkwork-rtc-sdk-go/rtcstandard/provider_package_loader.go
package rtcstandard

import "fmt"

type RtcProviderPackageLoaderError struct {
	Code    string
	Message string
}

func (e RtcProviderPackageLoaderError) Error() string {
	return fmt.Sprintf("%s: %s", e.Code, e.Message)
}

type RtcProviderPackageLoadRequest struct {
	ProviderKey     string
	PackageIdentity string
}

type RtcResolvedProviderPackageLoadTarget struct {
	PackageEntry RtcProviderPackageCatalogEntry
}

type RtcProviderModuleNamespace map[string]string
type RtcProviderPackageImportFn func(RtcResolvedProviderPackageLoadTarget) (RtcProviderModuleNamespace, error)
type RtcProviderPackageLoader func(RtcProviderPackageLoadRequest) (RtcProviderModuleNamespace, error)

type RtcProviderPackageInstallRequest struct {
	DriverManager any
	LoadRequest   RtcProviderPackageLoadRequest
}

func ResolveRtcProviderPackageLoadTarget(
	request RtcProviderPackageLoadRequest,
) (RtcResolvedProviderPackageLoadTarget, error) {
	var packageByProviderKey *RtcProviderPackageCatalogEntry
	if request.ProviderKey != "" {
		packageByProviderKey = GetRtcProviderPackageByProviderKey(request.ProviderKey)
	}

	var packageByIdentity *RtcProviderPackageCatalogEntry
	if request.PackageIdentity != "" {
		packageByIdentity = GetRtcProviderPackageByPackageIdentity(request.PackageIdentity)
	}

	if packageByProviderKey != nil && packageByIdentity != nil && packageByProviderKey.PackageIdentity != packageByIdentity.PackageIdentity {
		return RtcResolvedProviderPackageLoadTarget{}, RtcProviderPackageLoaderError{
			Code:    "provider_package_identity_mismatch",
			Message: "providerKey and packageIdentity must resolve to the same provider package boundary.",
		}
	}

	resolvedPackage := packageByProviderKey
	if resolvedPackage == nil {
		resolvedPackage = packageByIdentity
	}

	if resolvedPackage == nil {
		return RtcResolvedProviderPackageLoadTarget{}, RtcProviderPackageLoaderError{
			Code:    "provider_package_not_found",
			Message: "No official provider package matches the requested provider boundary.",
		}
	}

	return RtcResolvedProviderPackageLoadTarget{PackageEntry: *resolvedPackage}, nil
}

func CreateRtcProviderPackageLoader(importPackage RtcProviderPackageImportFn) RtcProviderPackageLoader {
	return func(request RtcProviderPackageLoadRequest) (RtcProviderModuleNamespace, error) {
		return LoadRtcProviderModule(request, importPackage)
	}
}

func LoadRtcProviderModule(
	request RtcProviderPackageLoadRequest,
	importPackage RtcProviderPackageImportFn,
) (RtcProviderModuleNamespace, error) {
	target, err := ResolveRtcProviderPackageLoadTarget(request)
	if err != nil {
		return nil, err
	}

	namespace, err := importPackage(target)
	if err != nil {
		return nil, RtcProviderPackageLoaderError{
			Code:    "provider_package_load_failed",
			Message: fmt.Sprintf("Reserved provider package loader scaffold could not load %s: %v", target.PackageEntry.PackageIdentity, err),
		}
	}

	if len(namespace) == 0 {
		return nil, RtcProviderPackageLoaderError{
			Code:    "provider_module_export_missing",
			Message: "Reserved provider package loader scaffold requires an executable provider module namespace.",
		}
	}

	return namespace, nil
}

func InstallRtcProviderPackage(
	request RtcProviderPackageInstallRequest,
	importPackage RtcProviderPackageImportFn,
) error {
	if _, err := LoadRtcProviderModule(request.LoadRequest, importPackage); err != nil {
		return err
	}

	return RtcProviderPackageLoaderError{
		Code:    "provider_package_load_failed",
		Message: "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
	}
}

func InstallRtcProviderPackages(
	requests []RtcProviderPackageInstallRequest,
	importPackage RtcProviderPackageImportFn,
) error {
	for _, request := range requests {
		if _, err := LoadRtcProviderModule(request.LoadRequest, importPackage); err != nil {
			return err
		}
	}

	if len(requests) > 0 {
		return RtcProviderPackageLoaderError{
			Code:    "provider_package_load_failed",
			Message: "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
		}
	}

	return nil
}

File: sdkwork-rtc-sdk-rust/src/provider_package_loader.rs
use std::collections::BTreeMap;

use crate::provider_package_catalog::{
    get_rtc_provider_package_by_package_identity, get_rtc_provider_package_by_provider_key,
    RtcProviderPackageCatalogEntry,
};

#[derive(Debug, Clone)]
pub struct RtcProviderPackageLoaderError {
    pub code: &'static str,
    pub message: String,
}

impl RtcProviderPackageLoaderError {
    pub fn new(code: &'static str, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

#[derive(Debug, Clone, Default)]
#[allow(non_snake_case)]
pub struct RtcProviderPackageLoadRequest {
    pub providerKey: Option<String>,
    pub packageIdentity: Option<String>,
}

#[allow(non_snake_case)]
pub struct RtcResolvedProviderPackageLoadTarget {
    pub packageEntry: &'static RtcProviderPackageCatalogEntry,
}

pub type RtcProviderModuleNamespace = BTreeMap<String, String>;
pub type RtcProviderPackageImportFn =
    fn(&RtcResolvedProviderPackageLoadTarget) -> Result<RtcProviderModuleNamespace, RtcProviderPackageLoaderError>;
pub type RtcProviderPackageLoader = Box<
    dyn Fn(RtcProviderPackageLoadRequest) -> Result<RtcProviderModuleNamespace, RtcProviderPackageLoaderError>,
>;

#[allow(non_snake_case)]
pub struct RtcProviderPackageInstallRequest<TDriverManager> {
    pub driverManager: TDriverManager,
    pub loadRequest: RtcProviderPackageLoadRequest,
}

pub fn resolve_rtc_provider_package_load_target(
    request: &RtcProviderPackageLoadRequest,
) -> Result<RtcResolvedProviderPackageLoadTarget, RtcProviderPackageLoaderError> {
    let package_by_provider_key = request
        .providerKey
        .as_deref()
        .and_then(get_rtc_provider_package_by_provider_key);
    let package_by_identity = request
        .packageIdentity
        .as_deref()
        .and_then(get_rtc_provider_package_by_package_identity);

    if let (Some(provider_key_entry), Some(package_identity_entry)) =
        (package_by_provider_key, package_by_identity)
    {
        if provider_key_entry.packageIdentity != package_identity_entry.packageIdentity {
            return Err(RtcProviderPackageLoaderError::new(
                "provider_package_identity_mismatch",
                "providerKey and packageIdentity must resolve to the same provider package boundary.",
            ));
        }
    }

    let resolved_package = package_by_provider_key.or(package_by_identity).ok_or_else(|| {
        RtcProviderPackageLoaderError::new(
            "provider_package_not_found",
            "No official provider package matches the requested provider boundary.",
        )
    })?;

    Ok(RtcResolvedProviderPackageLoadTarget {
        packageEntry: resolved_package,
    })
}

pub fn create_rtc_provider_package_loader(
    import_package: RtcProviderPackageImportFn,
) -> RtcProviderPackageLoader {
    Box::new(move |request| load_rtc_provider_module(&request, import_package))
}

pub fn load_rtc_provider_module(
    request: &RtcProviderPackageLoadRequest,
    import_package: RtcProviderPackageImportFn,
) -> Result<RtcProviderModuleNamespace, RtcProviderPackageLoaderError> {
    let target = resolve_rtc_provider_package_load_target(request)?;
    let namespace = import_package(&target).map_err(|error| {
        if error.code == "provider_package_load_failed" || error.code == "provider_module_export_missing" {
            error
        } else {
            RtcProviderPackageLoaderError::new(
                "provider_package_load_failed",
                format!(
                    "Reserved provider package loader scaffold could not load {}: {}",
                    target.packageEntry.packageIdentity, error.message
                ),
            )
        }
    })?;

    if namespace.is_empty() {
        return Err(RtcProviderPackageLoaderError::new(
            "provider_module_export_missing",
            "Reserved provider package loader scaffold requires an executable provider module namespace.",
        ));
    }

    Ok(namespace)
}

pub fn install_rtc_provider_package<TDriverManager>(
    request: &RtcProviderPackageInstallRequest<TDriverManager>,
    import_package: RtcProviderPackageImportFn,
) -> Result<(), RtcProviderPackageLoaderError> {
    let _namespace = load_rtc_provider_module(&request.loadRequest, import_package)?;

    Err(RtcProviderPackageLoaderError::new(
        "provider_package_load_failed",
        "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
    ))
}

pub fn install_rtc_provider_packages<TDriverManager>(
    requests: &[RtcProviderPackageInstallRequest<TDriverManager>],
    import_package: RtcProviderPackageImportFn,
) -> Result<(), RtcProviderPackageLoaderError> {
    for request in requests {
        let _namespace = load_rtc_provider_module(&request.loadRequest, import_package)?;
    }

    if !requests.is_empty() {
        return Err(RtcProviderPackageLoaderError::new(
            "provider_package_load_failed",
            "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands.",
        ));
    }

    Ok(())
}
